features/285-design-webhooks/events.md

```diff
@@ -0,0 +1,348 @@
+# Web-hook Events
+
+All web-hook events have the same format:
+```json
+{
+    "version": 1,
+    "id": "0E1E9537-DF4F-4AF6-8B48-3DB4574D4F24",
+    "seq": 435,
+    "type": "after_user_create",
+    "payload": {
+        // ...
+    },
+    "context": {
+        // ...
+    }
+}
+```
+
+- `version`: signed 32-bit integer, the version of the web-hook event; it is
+             specific to the event type.
+- `id`: string, ID of the web-hook event.
+- `seq`: signed 64-bit integer, monotonic increasing event sequence number to
+         establish event order.
+- `type`: string, the type of the web-hook event.
+- `payload`: object, the payload of the web-hook; it is specific to each event type.
+- `context`: object, the context of the web-hook.
+
+## Versioning
+The content of top-level fields `type`, `id` and are guaranteed to exist
+and only backward-compatible changes would be made.
+
+Field `version` in the payload would be updated when a backward incompatible
+change is made to `payload`, `context` or `seq`. Consumer should handle unknown
+version gracefully.
+
+## Context
+Each web-hook event would have a context, it contains the environmental
+information about the event:
+
+- `timestamp`:
+  timestamp of when this event is generated; retried deliveries would not
+  affect this field.
+- `request_id`:
+  ID of the HTTP request that generated this event; if the event is not
+  generated by HTTP request, this field would be `null`.
+- `user_id`:
+  the ID of user performing the request that generated this event; if request
+  is not authenticated or event is not generated by user action, this field
+  would be `null`.
+- `identity_id`:
+  the ID of identity performing the request that generated this event; if
+  request is not authenticated or event not generated by user action, this field
+  would be `null`.
+
+**NOTE**
+- When user is performing authentication (e.g. signing up/logging in),
+  the request will not be treated as authenticated with the user to
+  sign up/log in, in both BEFORE/AFTER events.
+
+
+## Mutations
+
+For BEFORE events about a user, following fields can be mutated:
```

Comment:
Can we also add `is_verified`?

Reply:
I'd prefer to add this when we provide a way to turn off auto-update of the flag.